runtime/include/map.h

```
@@ -0,0 +1,183 @@
+		                         .value     = malloc(value_len),
+		                         .value_len = value_len,
+		                         .next      = map->buckets[hash % MAP_BUCKET_COUNT] };
+
+	assert(new_node->key);
+	assert(new_node->value);
+
+	// Copy Key and Value
+	memcpy(new_node->key, key, key_len);
+	memcpy(new_node->value, value, value_len);
+
+	map->buckets[hash % MAP_BUCKET_COUNT] = new_node;
+	did_set                               = true;
+
+DONE:
+	LOCK_UNLOCK(&map->locks[hash % MAP_LOCK_STRIPES]);
+	return did_set;
+}
+
+/**
+ * @returns boolean if node was deleted or not
+ */
+static inline bool
+map_delete(struct map *map, uint8_t *key, uint32_t key_len)
+{
+	bool did_delete = false;
+
+	uint32_t hash = MAP_HASH(key, key_len);
+	LOCK_LOCK(&map->locks[hash % MAP_LOCK_STRIPES]);
+
+	struct map_node *prev = map->buckets[hash % MAP_BUCKET_COUNT];
+	if (prev->hash == hash) {
+		map->buckets[hash % MAP_BUCKET_COUNT] = prev->next;
+		free(prev->key);
+		free(prev->value);
+		free(prev);
+		did_delete = true;
+		goto DONE;
+	}
+
+	for (struct map_node *node = prev->next; node != NULL; prev = node, node = node->next) {
+		prev->next = node->next;
+		free(node->key);
+		free(node->value);
+		free(node);
+		did_delete = true;
+		goto DONE;
+	}
+
+DONE:
+	LOCK_UNLOCK(&map->locks[hash % MAP_LOCK_STRIPES]);
+	return did_delete;
+}
+
+static inline void
+map_upsert(struct map *map, uint8_t *key, uint32_t key_len, uint8_t *value, uint32_t value_len)
+{
+	uint32_t hash = MAP_HASH(key, key_len);
+	LOCK_LOCK(&map->locks[hash % MAP_LOCK_STRIPES]);
+	for (struct map_node *node = map->buckets[hash % MAP_BUCKET_COUNT]; node != NULL; node = node->next) {
+		if (node->hash == hash) {
+			node->value_len = value_len;
+			node->value     = realloc(node->value, value_len);
+			assert(node->value);
+			memcpy(node->value, value, value_len);
+		}
+		goto DONE;
+	}
+
+	struct map_node *new_node = (struct map_node *)malloc(sizeof(struct map_node));
```

Too late to care about this now, but in systems that don't want to check the return value for `malloc`, they often define a `xmalloc` (I don't know why `x`) that will panic out with an error message if `malloc` returns `NULL`. Better to fail early.

---

I added an `xmalloc` and used it in this file. Will at least be available for future use.

---

I added an `xmalloc` and used it in this file. Will at least be available for future use.